wurst/lib/ItemRecipe.wurst

```diff
@@ -0,0 +1,570 @@
+        if recipeMap.has(-itemTypeId)
+            return recipeMap.get(-itemTypeId)
+        return null
+
+    /** Returns first recipe which reward matches specified item type. */
+    static function getRecipe(int itemTypeId) returns ItemRecipe
+        let recipes = getRecipes(itemTypeId)
+        if recipes != null
+            return recipes.getFirst()
+        return null
+
+    /** Returns recipes that specified item is part of. */
+    static function getRecipesForIngredient(int itemTypeId) returns LinkedList<ItemRecipe>
+        if recipeMap.has(itemTypeId)
+            return recipeMap.get(itemTypeId)
+        return null
+
+    /** Returns recipes that can be assembled by casting specified ability. */
+    static function getRecipesForAbility(int abilityId) returns LinkedList<ItemRecipe>
+        if recipeMap.has(abilityId)
+            return recipeMap.get(abilityId)
+        return null
+
+    /** Starts single-reference counted batch. Allows to assign multiple items to the same item slot. */
+    function startBatch()
+        if not batch
+            batch = true
+        // TODO debug message
+
+    /** Closes current batch. */
+    function endBatch()
+        if batch
+            batch = false
+            count++
+
+    /** Sets or removes specified ability from triggering recipe assembly. */
+    function setAbility(int abilityId)
+        LinkedList<ItemRecipe> recipes
+
+        if this.abilityId == abilityId
+            return
+
+        if this.abilityId != 0
+            recipes = recipeMap.get(abilityId)
+            recipes.remove(this)
+            if recipes.isEmpty()
+                destroy recipes
+                recipeMap.remove(abilityId)
+
+        if abilityId > 0
+            this.abilityId = abilityId
+            if not recipeMap.has(abilityId)
+                recipes = new LinkedList<ItemRecipe>()
+                recipes.push(this)
+                recipeMap.put(abilityId, recipes)
+            else
+                recipes = recipeMap.get(abilityId)
+                if not recipes.contains(this)
+                    recipes.push(this)
+
+    /** Removes all entries that match specified item type from recipe ingredient list. */
+    function removeItem(int itemTypeId)
+        let iter = ingredients.iterator()
+        boolean found = false
+
+        while iter.hasNext()
+            if iter.next().itemTypeId == itemTypeId
+                iter.remove()
+                found = true
+        iter.close()
+
+        if found
+            let recipes = recipeMap.get(itemTypeId)
+            recipes.remove(this)
+            if recipes.isEmpty()
+                destroy recipes
+                recipeMap.remove(itemTypeId)
+
+    /** Adds new entry to recipe ingredient list. */
+    function addItem(int itemTypeId, boolean remove, int charges)
+        LinkedList<ItemRecipe> recipes
+        int chrgs = charges
+
+        if itemTypeId <= 0 or itemTypeId == reward
+            return
```

Comment:
Let's also show an error if someone tries to do this

Reply:
Debug messages are not part of current implementation. They will be added later. I don't want to add random message here and there, I want it to be organized - and this takes time.